### PR TITLE
[SYCL][E2E] Replace TEMPORARY_DISABLED by UNSUPPORTED

### PR DESCRIPTION
**Pull request by KornevNikita**
To unify the mechanism to disable tests this patch replaces `REQUIRES: TEMPORARY_DISABLED` by `UNSUPPORTED: true` to disable the test. Also updating the tests using "TEMPORARY_DISABLED".

### CODE REVIEW

File: sycl/test-e2e/Basic/free_function_queries/free_function_queries.cpp
Comment (sarnex) [inline sycl/test-e2e/Basic/free_function_queries/free_function_queries.cpp:2]: maybe a nitpick but `overall` is a bit unclear to me, maybe like `will_never_work` or `never_expected_to_work`, or maybe something shorter i can't think of

Comment (KornevNikita) [reply sycl/test-e2e/Basic/free_function_queries/free_function_queries.cpp:2]: I thought about something this, but what if it will work someday? :)

Comment (sarnex) [reply sycl/test-e2e/Basic/free_function_queries/free_function_queries.cpp:2]: i guess we need the code owners to decide that, maybe for now we can do like `UNSUPPORTED: temporary` and ask code reviewers to update it to `permanent` or `never_will_work` or whatever?

Comment (sarnex) [reply sycl/test-e2e/Basic/free_function_queries/free_function_queries.cpp:2]: my assumption was that it's supported on some configurations but not others (will always fail on cuda, should work elsewhere). if the test is unsupported on all platforms and the code owner doesn't expect the change, i agree we should remove the test.

File: sycl/test-e2e/README.md
Comment (bader) [inline sycl/test-e2e/README.md:379]: Please, add clarification, why and when `REQUIRES` should be avoided. LLVM's testing guide has following tips: https://llvm.org/docs/TestingGuide.html#tips-for-writing-constraints.

**Comment by bader** [comment]
@KornevNikita, I'm not sure I understand the problem you are trying to solve with this change. 

> the disabled test may be missed by grep

Did I get it right that you are using grep to find disabled tests? If so, please, try using `llvm-lit` instead of `grep` tool. Here are useful command line options:

`--show-excluded` Show excluded tests.
`--show-skipped` Show skipped tests.
`--show-unsupported` Show unsupported tests.
`--show-used-features` Show all features used in the test suite (in XFAIL, UNSUPPORTED and REQUIRES) and exit.

Full list of options: https://llvm.org/docs/CommandGuide/lit.html

File: sycl/test-e2e/SpecConstants/2020/non_native/gpu.cpp
Comment (bader) [inline sycl/test-e2e/SpecConstants/2020/non_native/gpu.cpp:1]:
```suggestion
// UNSUPPORTED: *
```
Why can't we use standard idiom?

**Comment by KornevNikita** [comment]
> @KornevNikita, I'm not sure I understand the problem you are trying to solve with this change.
> 
> > the disabled test may be missed by grep
> 
> Did I get it right that you are using grep to find disabled tests? If so, please, try using `llvm-lit` instead of `grep` tool. Here are useful command line options:
> 
> `--show-excluded` Show excluded tests. `--show-skipped` Show skipped tests. `--show-unsupported` Show unsupported tests. `--show-used-features` Show all features used in the test suite (in XFAIL, UNSUPPORTED and REQUIRES) and exit.
> 
> Full list of options: https://llvm.org/docs/CommandGuide/lit.html

The main idea is to unify the mechanism to disable tests, as this "TEMPORARY_DISABLED" is non-standard, but some work-around. We could just switch it to `UNSUPPORTED: *` - but it doesn't accept the asterisk.

**Comment by bader** [comment]
> > @KornevNikita, I'm not sure I understand the problem you are trying to solve with this change.
> > > the disabled test may be missed by grep
> > 
> > 
> > Did I get it right that you are using grep to find disabled tests? If so, please, try using `llvm-lit` instead of `grep` tool. Here are useful command line options:
> > `--show-excluded` Show excluded tests. `--show-skipped` Show skipped tests. `--show-unsupported` Show unsupported tests. `--show-used-features` Show all features used in the test suite (in XFAIL, UNSUPPORTED and REQUIRES) and exit.
> > Full list of options: https://llvm.org/docs/CommandGuide/lit.html
> 
> The main idea is to unify the mechanism to disable tests, as this "TEMPORARY_DISABLED" is non-standard, but some work-around. We could just switch it to `UNSUPPORTED: *` - but it doesn't accept the asterisk.

From my POV, replacing a single way (i.e. `REQUIRES: TEMPORARY_DISABLED`) to disable a LIT test with 3 new ways (i.e. `UNSUPPORTED: overall, flaky, hangs`) doesn't make it standard. It's still a temporary work-around of some test issues.

> We could just switch it to `UNSUPPORTED: *` - but it doesn't accept the asterisk.

Sorry, it looks like the right way to use `UNSUPPORTED: true`.

**Comment by KornevNikita** [comment]
> Sorry, it looks like the right way to use UNSUPPORTED: true.
Didn't think about that. If this works I'd prefer to proceed with it.

**Comment by bader** [comment]
> > Sorry, it looks like the right way to use UNSUPPORTED: true.
> 
> Didn't think about that. If this works I'd prefer to proceed with it.

It's used by a few tests in llvm-project.

compiler-rt/test/builtins/Unit/gcc_personality_test.c

> // FIXME: UNSUPPORTED as currently it cannot be built by lit properly.
> // UNSUPPORTED: true

Other uses are in `bolt/test/X86/reader-stale-yaml-std.test` and `bolt/test/X86/dwarf4-ftypes-dwp-input-dwp-output.test`.

**Comment by KornevNikita** [comment]
@bader thanks for pointing out! I'll update the PR

File: sycl/test-e2e/SpecConstants/2020/non_native/gpu.cpp
Comment (KornevNikita) [reply sycl/test-e2e/SpecConstants/2020/non_native/gpu.cpp]: As discussed below moved to `true` - [67f66ab](https://github.com/intel/llvm/pull/15946/commits/67f66ab55076ead0e2325f08e799967d305d8c2f).

File: sycl/test-e2e/README.md
Comment (bader) [inline sycl/test-e2e/README.md:351]:
```suggestion
  backend / device / OS / etc.
```

File: sycl/test-e2e/README.md
Comment (AlexeySachkov) [reply sycl/test-e2e/README.md]: > Generally, people use `REQUIRES` to state things that the test depends on to operate correctly, and `UNSUPPORTED` to exclude cases where the test is expected never to work.

`REQUIRES: TEMPORARY_DISABLED` is explicitly the case where the test is expected never to work, so `UNSUPPORTED` should be used.

But our main motivation is to switch to a mechanism which we track: `REQUIRES: TEMPORARY_DISABLED` can easily go without any trackers being created, leaving that test disabled **for years**:
- `sycl/test-e2e/Basic/image/image_accessor_range.cpp` - has been disabled for 2 years
- `sycl/test-e2e/Regression/complex_global_object.cpp` - has been disabled for at least 4 years

We have a mechanism to track that every `UNSUPPORTED` directive either has an explanation about why is it intended, or has a tracker associated with it, see https://github.com/intel/llvm/blob/sycl/sycl/test/e2e_test_requirements/no-unsupported-without-info.cpp

File: sycl/test-e2e/README.md
Comment (KornevNikita) [reply sycl/test-e2e/README.md]: [faaf532](https://github.com/intel/llvm/pull/15946/commits/faaf532f36046e698c0d8158dfbf21a2a76266a9)

File: sycl/test-e2e/README.md
Comment (bader) [reply sycl/test-e2e/README.md]: I'm not sure I understand the problem with tracking tests with `REQUIRES` instead of `UNSUPPORTED`. How are you going to control which keyword is used? Don't we replace one problem with another?

Comment (bader) [reply sycl/test-e2e/README.md]: I'm not convinced that we have to create artificial limitations for developers to enable the tracking. I believe that we can catch misuses of `REQUIRES` keyword using the options I mentioned in [this comment](https://github.com/intel/llvm/pull/15946#issuecomment-2455482606).

Comment (KornevNikita) [reply sycl/test-e2e/README.md]: Just interesting - I plan to add a test to check if all features passed to `REQUIRES` exist. Will it be then a limitation to prohibit developers to put there everything they want?

Comment (KornevNikita) [reply sycl/test-e2e/README.md]: Asking because TEMPORARY_DISABLED is not documented anywhere (or at least I didn't see it)

Comment (AlexeySachkov) [reply sycl/test-e2e/README.md]: > I'm not convinced that we have to create artificial limitations for developers to enable the tracking. I believe that we can catch misuses of `REQUIRES` keyword using the options I mentioned in [this comment](https://github.com/intel/llvm/pull/15946#issuecomment-2455482606).

So, we do use them (at least some of them) in pre-commit and I do not see how it helps. Having a developer to go look into their pre-commit and make sure that their new/modified test are not listed as skipped/unsupported is even less reliable than code review.

Moreover, to detect tests which are **already** accidentally skipped, we would have to review them all manually. Or write a script I was talking about, which makes us one step away from enabling it as a test.

Considering that the situation when a test is completely disabled everywhere is relatively rare (we only have ~10 tests like this in the whole 1000+ tests codebase), I do not think that any limitations we are imposing are severe.

Comment (bader) [reply sycl/test-e2e/README.md]: > Just interesting - I plan to add a test to check if all features passed to `REQUIRES` exist. Will it be then a limitation to prohibit developers to put there everything they want?

I think this is want we want.

**Review by againull — APPROVED** [review]

File: sycl/test-e2e/README.md
Comment (steffenlarsen) [reply sycl/test-e2e/README.md]: Just to try and converge the discussion here, the intention of this line in the README is to consolidate the ways we disable tests. Previously, we would have `UNSUPPORTED` in some cases, `XFAIL` in others, and `REQUIRES: TEMPORARILY_DISABLED` in a few. Now we want to unify these, which we do by:

1. Using `UNSUPPORTED` for cases that are intentionally unsupported for a given feature, e.g. using Intel device only features and disabling for CUDA and HIP. Note that `REQUIRES` is also an option for these, but sometimes exclusion is better than inclusion.
2. Using `UNSUPPORTED` for cases where the tests are currently unintentionally flakily failing or hangs happen.
3. Using `XFAIL` for cases where tests are unintentionally failing, excluding the cases from the previous point.

That leaves `REQUIRES` to be used exclusively for intentional requirements. From a tracking and code-quality POV, I argue this comes with the following benefits:

- Developers are required to add JIRA and/or Github trackers when disabling tests, which helps us make sure that these failures don't go uninvestigated for a _long_ time, as we have seen happen before.
- Gathering statistics on unintentionally failing tests is a lot easier.
- It allows us to make a script that can check for known features in `REQUIRES` fields to avoid cases where a misspelling of a feature unintentionally disables a test on all targets. Technically, we could include `TEMPORARILY_DISABLED` as a "known" feature, but I see no point in doing so when we have alternatives that should be used instead, given the two points above.

@bader - Do I understand correctly that the initial point of this comment thread was to refine the wording here? If so, would it be something like:

> Please do not use `REQUIRES` to temporarily disable a given test on targets that unintentionally fail it. Use `XFAIL` or `UNSUPPORTED` instead.

Comment (KornevNikita) [reply sycl/test-e2e/README.md]: @bader do you still think we shouldn't merge this one?

Comment (bader) [reply sycl/test-e2e/README.md]: > @bader - Do I understand correctly that the initial point of this comment thread was to refine the wording here?

Yes.

> If so, would it be something like:
> 
> > Please do not use `REQUIRES` to temporarily disable a given test on targets that unintentionally fail it. Use `XFAIL` or `UNSUPPORTED` instead.

This wording addresses only `when` part of my request, but not `why`. Based on the answers in this thread, it sounds like the reason to avoid `REQUIRES` is that it's easier to misuse then `UNSUPPORTED`. Did I get it right?

Comment (bader) [reply sycl/test-e2e/README.md]: > @bader do you still think we shouldn't merge this one?

I don't object to merging, but I have doubts that this PR eliminates the root cause of the problem with today's using `REQUIRES`. I believe there are better ways to track if there are misuses of llvm-lit keywords.

@KornevNikita, please, resolve merge conflicts before merging.

Comment (KornevNikita) [reply sycl/test-e2e/README.md]: > Based on the answers in this thread, it sounds like the reason to avoid REQUIRES is that it's easier to misuse then UNSUPPORTED. Did I get it right?

As one of the reasons. I'd also say that `TEMPORARY_DISABLED` is some kind of work-around. And we have a more official alternative in the form of `UNSUPPORTED`, which we are also already tracking.

Comment (bader) [reply sycl/test-e2e/README.md]: > > Based on the answers in this thread, it sounds like the reason to avoid REQUIRES is that it's easier to misuse then UNSUPPORTED. Did I get it right?
> 
> As one of the reasons. I'd also say that `TEMPORARY_DISABLED` is some kind of work-around. And we have a more official alternative in the form of `UNSUPPORTED`, which we are also already tracking.

I don't agree with that statement. Both are viable options. I guess you call the use of `UNSUPPORTED` keyword "official" because it's documented. We could make use of `REQUIRES` official as well if we wanted. 

I'm okay with the idea of setting the convention to use `UNSUPPORTED` for SYCL tests disabling.

**Review by AlexeySachkov — APPROVED** [review]

**Comment by KornevNikita** [comment]
@intel/dpcpp-esimd-reviewers could you please take a look?

**Review by sarnex — APPROVED** [review]
sorry, was OOO. lgtm!

**Pull request merged**
